Subject: [PATCH] Fix crash issue on Storage Manager app

StorageManager app Start/Exit many times occur crash if there are over
 100 unused app.
Because ApplicationsState.BackgroundHandler will update AppEntry'Size
during the ApplicationsState.MainHandler sort all AppEntry with
SIZE_COMPARATOR. In this case, SIZE_COMPARATOR will occur
"IllegalArgumentException Comparison method violates its general
contract!".
In ordre to fix the crash, add mEntriesMap synchronized during the AppEntry
sort operation.

diff --git a/packages/SettingsLib/src/com/android/settingslib/applications/ApplicationsState.java b/packages/SettingsLib/src/com/android/settingslib/applications/ApplicationsState.java
index f0ec107..a803e60 100644
--- a/packages/SettingsLib/src/com/android/settingslib/applications/ApplicationsState.java
+++ b/packages/SettingsLib/src/com/android/settingslib/applications/ApplicationsState.java
@@ -649,7 +649,11 @@
             }
 
             if (comparator != null) {
-                Collections.sort(filteredApps, comparator);
+                synchronized (mEntriesMap) {
+                    // Locking to ensure that the background handler does not mutate
+                    // the size of AppEntries used for ordering while sorting.
+                    Collections.sort(filteredApps, comparator);
+                }
             }
 
             synchronized (mRebuildSync) {
